### DISCUSSION TITLE
Websocket flow control

### DISCUSSION THREAD

**Discussion by mkiefel** [foxglove/community#744] [Q&A]
Hi! First, thanks for this amazing piece of software!

I recently implemented a subset of the Foxglove ws protocol in Rust for our project. Something that I noticed is, that if I hammered Foxglove Studio with a little too much data, e.g., an uncompressed image stream + a point cloud with ~15Hz, such that traffic would be sightly north of 500Mbit/s, that Foxglove Studio would start to lag and consume more and more memory.

I checked the sending side, and I believe that stuff was leaving the server implementation.

Reading up a little on it [1], I believe that this might be due to missing flow control for websockets if the consumer is not able to keep up in the browser. I would be interested if you believe that there is value in adding some kind of application level flow-control built into the protocol to inform the sender to stop sending/drop messages.

[1] https://stackoverflow.com/a/58029590

**Comment by mkiefel** [comment]
For the curious: I've now published the code in https://github.com/mkiefel/foxglove-ws.

**Reply by jtbandes** [reply]
Very cool! I think if this implementation stabilizes and you are interested in making a pull request to the official ws-protocol repo, it would be nice to have a rust crate for this.

Application-level flow control is probably not out of the question, but we would want to understand more about what is actually happening in the situation you describe. It's possible that the app could be improved to avoid buffering indefinitely when it lags behind on rendering. It's also possible that some browser limitations prevent us from making use of tcp's native flow control and application-level control would be needed. Unfortunately the ideal corrective action differs based on the visualization – it might be okay to drop point cloud messages sometimes, but if the user has configured a decay time then they're expecting to see a history of point clouds and not just the latest ones.

**Reply by mkiefel** [reply]
Sounds great thanks.

Thanks also for the context!
